RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc

```diff
@@ -0,0 +1,637 @@
+    double charge = channelData.tsRawCharge(iTS);
+    double ped = channelData.tsPedestal(iTS);
+
+    nnlsWork_.amplitudes.coeffRef(iTS) = charge - ped;
+
+    //ADC granularity
+    double noiseADC = (1./sqrt(12))*channelData.tsDFcPerADC(iTS);
+
+    //Dark current (for SiPMs)
+    double noiseDC=0;
+    if(channelData.hasTimeInfo() && !channelData.hasEffectivePedestals() && (charge-ped)>channelData.tsPedestalWidth(iTS)) {
+      noiseDC = darkCurrent;
+    }
+
+    //Photostatistics
+    double noisePhoto = 0;
+    if ( (charge-ped)>channelData.tsPedestalWidth(iTS)) {
+      noisePhoto = sqrt((charge-ped)*channelData.fcByPE());
+    }
+
+    //Electronic pedestal
+    double pedWidth = channelData.tsPedestalWidth(iTS);
+
+    //Total uncertainty from all sources
+    nnlsWork_.noiseTerms.coeffRef(iTS) = noiseADC*noiseADC + noiseDC*noiseDC + noisePhoto*noisePhoto + pedWidth*pedWidth;
+
+    tsTOT += charge - ped;
+    if( iTS==nnlsWork_.tsOffset ){
+      tstrig += (charge - ped)*channelData.tsGain(0);
+    }
+  }
+
+  if(tstrig >= ts4Thresh_) {
+
+    useTriple=false;
+
+    // only do pre-fit with 1 pulse if chiSq threshold is positive
+    if (chiSqSwitch_>0) {
+      doFit(reconstructedVals,1);
+      if (reconstructedVals[1]>chiSqSwitch_) {
+	doFit(reconstructedVals,0); //nbx=0 means use configured BXs
+	useTriple=true;
+      }
+    }
+    else {
+      doFit(reconstructedVals,0);
+      useTriple=true;
+    }
+  }
+  else{
+    reconstructedVals.clear();
+    reconstructedVals.push_back(0.);
+    reconstructedVals.push_back(888.);
+    reconstructedVals.push_back(888.);
+  }
+  
+  reconstructedEnergy = reconstructedVals[0]*channelData.tsGain(0);
+  reconstructedTime = reconstructedVals[1];
+  chi2 = reconstructedVals[2];
+
+}
+
+void MahiFit::doFit(std::vector<float> &correctedOutput, int nbx) const {
+
+  unsigned int bxSize=1;
+
+  if (nbx==1) {
+    nnlsWork_.bxOffset = 0;
+  }
+  else {
+    bxSize = bxSizeConf_;
+    nnlsWork_.bxOffset = bxOffsetConf_;
+  }
+
+  nnlsWork_.bxs.resize(bxSize);
+
+  if (nbx==1) {
+    nnlsWork_.bxs.coeffRef(0) = 0;
+  }
+  else {
+    for (unsigned int iBX=0; iBX<bxSize; iBX++) {
+      nnlsWork_.bxs.coeffRef(iBX) = activeBXs_[iBX];
+    }
+  }
+
+  nnlsWork_.nPulseTot = bxSize;
+
+  if (dynamicPed_) {
+    nnlsWork_.nPulseTot++;
+    nnlsWork_.bxs.resize(nnlsWork_.nPulseTot);
+    nnlsWork_.bxs[nnlsWork_.nPulseTot-1] = 100;
+  }
+
+  nnlsWork_.pulseMat.resize(nnlsWork_.tsSize,nnlsWork_.nPulseTot);
+  nnlsWork_.ampVec = PulseVector::Zero(nnlsWork_.nPulseTot);
+  nnlsWork_.errVec = PulseVector::Zero(nnlsWork_.nPulseTot);
+
+  int offset=0;
+  for (unsigned int iBX=0; iBX<nnlsWork_.nPulseTot; iBX++) {
+    offset=nnlsWork_.bxs.coeff(iBX);
+
+    nnlsWork_.pulseShapeArray[iBX] = FullSampleVector::Zero(MaxFSVSize);
+    nnlsWork_.pulseDerivArray[iBX] = FullSampleVector::Zero(MaxFSVSize);
+    nnlsWork_.pulseCovArray[iBX]   = FullSampleMatrix::Constant(0);
+
+    if (offset==100) {
```

Comment:
please define a named constexpr (e.g. ```uninitializedOffset = 100```) and use it, instead of the magic 100 used in 3 places